plutus-tx/src/PlutusTx/Show/TH.hs

```diff
@@ -46,5 +77,194 @@ concatBuiltinStrings = \case
         let (ys, zs) = splitAt (length xs `divideInteger` 2) xs
          in concatBuiltinStrings ys `appendString` concatBuiltinStrings zs
 
------------------------------------------------------------
--- TODO: TH machinery for deriving `Show` here
+-- | Derive `Show` instance. Adapted from @Text.Show.Deriving.deriveShow@.
+deriveShow :: TH.Name -> TH.DecsQ
+deriveShow name = do
+    TH.DatatypeInfo
+        { TH.datatypeName = tyConName
+        , TH.datatypeInstTypes = tyVars0
+        , TH.datatypeCons = cons
+        } <-
+        TH.reifyDatatype name
+    let -- The purpose of the `TH.VarT . varTToName` roundtrip is to remove the kind
+        -- signatures attached to the type variables in `tyVars0`. Otherwise, the
+        -- `KindSignatures` extension would be needed whenever `length tyVars0 > 0`.
+        tyVars = TH.VarT . varTToName <$> tyVars0
+        instanceCxt :: TH.Cxt
+        instanceCxt = TH.AppT (TH.ConT ''Show) <$> tyVars
+        instanceType :: TH.Type
+        instanceType = TH.AppT (TH.ConT ''Show) $ foldl' TH.AppT (TH.ConT tyConName) tyVars
+        showsPrecDecs = deriveShowsPrec cons
+    pure <$> TH.instanceD (pure instanceCxt) (pure instanceType) showsPrecDecs
+
+-- | Derive `showsPrec` definition for each data constructor.
+deriveShowsPrec :: [TH.ConstructorInfo] -> [TH.DecQ]
+deriveShowsPrec cons =
+    [ TH.funD 'showsPrec [clause]
+    , -- `showsPrec` must be inlinable for the plugin to inline it
+      TH.pragInlD 'showsPrec TH.Inlinable TH.FunLike TH.AllPhases
+    ]
+  where
+    clause = TH.clause [] body []
+    body = TH.normalB $ deriveShowsPrecBody cons
+
+deriveShowsPrecBody :: [TH.ConstructorInfo] -> TH.ExpQ
+deriveShowsPrecBody cons = do
+    p <- TH.newName "_p" -- The precedence argument. It is not always used, hence the leading `_`.
+    value <- TH.newName "_value" -- The value to be shown
+    let pats = [TH.varP p, TH.varP value]
+        body = TH.caseE (TH.varE value) (deriveMatchForCon p <$> cons)
+    TH.lamE pats body
+
+{- | Derive `showsPrec` body for a single data constructor.
+
+ Example: given `C Integer Bool`, this generates
+
+ @
+    case _value of C arg1 arg2 ->
+        showParen
+            (_p `greaterThanInteger` 10)
+            (showString "C " . showsPrec 11 arg1 . showSpace . showsPrec 11 arg2)
+ @
+-}
+deriveMatchForCon :: TH.Name -> TH.ConstructorInfo -> TH.MatchQ
+deriveMatchForCon p = \case
+    TH.ConstructorInfo
+        { TH.constructorName = conName
+        , TH.constructorFields = []
+        } ->
+            TH.match
```

Comment:
It's slightly better to use a quote, but slightly worse because all other cases also need to construct the entire `case`, so I'll call it a wash :smile: